pandas/tests/test_timedelta64_overflow.py

```diff
@@ -0,0 +1,322 @@
+    }
+    return_type = timestamp_types[max(type_hierarchy[t] for t in request.param)]
+
+    return BinaryOpTypes(request.param[0], request.param[1], return_type)
+
+
+def wrap_value(value: Union[Timestamp, Timedelta], type_):
+    """
+    Return value wrapped in a container of given type_, or as-is if type_ is a scalar.
+    """
+    if issubclass(type_, (Timedelta, Timestamp)):
+        return type_(value)
+
+    if issubclass(type_, ExtensionArray):
+        box_cls = array
+    elif issubclass(type_, Index):
+        box_cls = Index
+    else:
+        box_cls = type_
+
+    return type_(tm.box_expected([value], box_cls))
+
+
+@given(positive_td=positive_tds)
+def test_add_raises_expected_error_if_result_would_overflow(
+    add_sub_types: BinaryOpTypes,
+    positive_td: Timedelta,
+):
+    left = wrap_value(Timedelta.max, add_sub_types.left)
+    right = wrap_value(positive_td, add_sub_types.right)
+
+    if add_sub_types.result is Timedelta:
+        msg = "|".join(
+            [
+                "int too big to convert",
+                "Python int too large to convert to C long",
+            ]
+        )
+    else:
+        msg = "Overflow in int64 addition"
+
+    with pytest.raises(OverflowError, match=msg):
+        left + right
+
+    with pytest.raises(OverflowError, match=msg):
+        right + left
+
+
+@xfail_no_overflow_check
+@given(positive_td=positive_tds)
+def test_sub_raises_expected_error_if_result_would_overflow(
+    add_sub_types: BinaryOpTypes,
+    positive_td: Timedelta,
+):
+    left = wrap_value(Timedelta.min, add_sub_types.left)
+    right = wrap_value(positive_td, add_sub_types.right)
+
+    msg = "Overflow in int64 addition"
+    with pytest.raises(OverflowError, match=msg):
+        left - right
+
+    with pytest.raises(OverflowError, match=msg):
+        (-1 * right) - abs(left)
+
+
+@given(td_value=positive_tds)
+def test_add_timestamp_raises_expected_error_if_result_would_overflow(
+    ts_add_sub_types: BinaryOpTypes,
+    td_value: Timedelta,
+):
+    left = wrap_value(td_value, ts_add_sub_types.left)
+    right = wrap_value(Timestamp.max, ts_add_sub_types.right)
+
+    ex = (OutOfBoundsDatetime, OverflowError)
+    msg = "|".join(["Out of bounds nanosecond timestamp", "Overflow in int64 addition"])
+
+    with pytest.raises(ex, match=msg):
+        left + right
+
+    with pytest.raises(ex, match=msg):
+        right + left
+
+
+@xfail_no_overflow_check
+@given(td_value=positive_tds)
+def test_sub_timestamp_raises_expected_error_if_result_would_overflow(
+    ts_add_sub_types: BinaryOpTypes,
+    td_value: Timedelta,
+):
+    right = wrap_value(td_value, ts_add_sub_types[0])
+    left = wrap_value(Timestamp.min, ts_add_sub_types[1])
+
+    ex = (OutOfBoundsDatetime, OverflowError)
+    msg = "|".join(["Out of bounds nanosecond timestamp", "Overflow in int64 addition"])
+
+    with pytest.raises(ex, match=msg):
+        left - right
+
+
+@given(value=st.floats().filter(lambda f: abs(f) > 1))
```

Comment:
is this really an improvement over e.g. `value = np.random.random()**2`?  I don't have a strong opinion, mostly curious.  how much test-runtime overhead does using hypothesis for this add?

Reply:
Heh, unsure, but also wonder the same. For simplicity, I switched to a few hardcoded values for now, but am curious to look into this more in the future (or hear from someone who has).